File: javascript/section03/bigORule2.js
//Drop the constants

function printFirstItemThenFirstHalfThenSayHi100Times(items) {
    console.log('thats me: ',items[0]); //O(1)

    var middleIndex = Math.floor(items.length /2); //O(n/2)
    var index = 0; //0(1)

    while (index < middleIndex) { //O(n/2)
           console.log(items[index]); 
           index++; 
    }

    for (var i = 0; i < 100; i++){  //O(100)
        console.log('hi');
    }
}


let array = new Array(100).fill('hey');


printFirstItemThenFirstHalfThenSayHi100Times(array);


/* 
    O(1 + n/2  + 100)
    O(n + 1)
    O(n)

*/


const compressBoxesTwice = (boxes) => {
    boxes.forEach(box => {
        console.log(box);
    });

    boxes.forEach(box => {
        console.log(box);
    });


    /* Because we have forEach in two different scopes we add them 
        So the BigO would be: O(2n)

        n for each forEach n + n = 2n;

        If we drop the constants it would be O(n)

    */
}

File: cpp/section03/bigORule2.cpp
#include <iostream>
#include <vector>

// Function to print the first item, then the first half, and say "hi" 100 times
void printFirstItemThenFirstHalfThenSayHi100Times(const std::vector<std::string>& items) {
    // O(1) - Print the first item
    std::cout << "That's me: " << items[0] << std::endl;

    // O(n/2) - Print the first half of the array
    int middleIndex = items.size() / 2; // O(1)
    int index = 0; // O(1)

    while (index < middleIndex) { // O(n/2)
        std::cout << items[index] << std::endl;
        index++;
    }

    // O(100) - Say "hi" 100 times
    for (int i = 0; i < 100; i++) { // O(100)
        std::cout << "hi" << std::endl;
    }
}

void compressBoxesTwice(const std::vector<std::string>& boxes) {
    // O(n) - Print each box once
    for (const auto& box : boxes) {
        std::cout << box << std::endl;
    }

    // O(n) - Print each box again
    for (const auto& box : boxes) {
        std::cout << box << std::endl;
    }

    // O(2n) -> O(n) after dropping constants
}

int main() {
    std::vector<std::string> array(100, "hey");

    // Call the first function
    printFirstItemThenFirstHalfThenSayHi100Times(array);

    // Call the compressBoxesTwice function
    compressBoxesTwice(array);

    return 0;
}

/*
    For the `printFirstItemThenFirstHalfThenSayHi100Times` function:
    O(1 + n/2 + 100) simplifies to O(n).

    For the `compressBoxesTwice` function:
    O(2n) simplifies to O(n) after dropping constants.
*/
